### ISSUE TITLE
Bound node invariants violated in scripting

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#25702]
I am adding code to check some invariants of the bound nodes.

One invariant is that every local variable that is used in the bound tree somewhere is declared (in the list of declared locals of a bound node) somewhere enclosing the use.

Three tests violate this, reporting
``` none
Test 'Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen.GotoTests.AcrossScriptDeclarations' failed: Microsoft.CodeAnalysis.ThrowingTraceListener+DebugAssertFailureException : undeclared local Local System.Int32 F

Test 'Microsoft.CodeAnalysis.CSharp.UnitTests.PatternMatchingTests_Global.GlobalCode_LabeledStatement_06' failed: Microsoft.CodeAnalysis.ThrowingTraceListener+DebugAssertFailureException : undeclared local Local System.Boolean d

Test 'Microsoft.CodeAnalysis.CSharp.UnitTests.OutVarTests.GlobalCode_LabeledStatement_06' failed: Microsoft.CodeAnalysis.ThrowingTraceListener+DebugAssertFailureException : undeclared local Local System.Boolean d
```

In each case there is a script variable by that name declared at the top level, and the bound tree contains a declaration for a local of that name, but does not declare it. In each case the code contains some error as well.

These tests are being skipped for now.

**Comment by gafter** [comment]
There are also around 50 tests that violate this in `CSharpExpressionCompilerTest`.

**Comment by tmat** [comment]
I don't think we should disable any tests. This will lead to regressions because we are not gonna fix these tests any time soon.

**Comment by gafter** [comment]
They are skipped in a PR that is targeting the branch `features/recursive-patterns`. I expect there is plenty of time to fix this bug before the code is ultimately integrated, in which case the tests won't ever need to be skipped in `master`. Of course this presumes the bug is fixed.

**Comment by gafter** [comment]
Marking this issue for P2 so the interactive team has a chance to fix the underlying bug before pattern-matching is intergrated into master, which will likely occur within the next couple of weeks.

**Comment by gafter** [comment]
Most of these tests are now enabled under Release builds.  Others fail because our stack optimizer requires the declarations, and those remain skipped.
